tests/utils.py

```diff
@@ -221,3 +241,212 @@ def integrate_average_end_points(values: pandas.Series) -> float:
+            import numpy as np
+
+            return np.array([self._random.randint(1, 100) for _ in range(3)], dtype=np.int64)
+        elif annotation == "NumpyDatetime64NSArray":
+            import numpy as np
+
+            return np.array([self._random.randint(1, 1704067200000) for _ in range(3)], dtype="datetime64[ms]")
+
+        resource_module_vars = vars(importlib.import_module(resource_cls.__module__))
+        annotation_cls = resource_module_vars.get(annotation)
+        is_class = inspect.isclass(annotation_cls)
+        if get_origin(annotation_cls) is not None:
+            annotation = self._get_typing_args(annotation_cls)
+            return self._create_complex_value(name, annotation, resource_cls)
+        elif issubclass(resource_cls, CogniteResourceList):
+            return [self._create(resource_cls._RESOURCE) for _ in range(self._random.randint(1, 3))]
+        elif isinstance(annotation_cls, enum.EnumMeta):
+            return self._random.choice(list(annotation_cls))
+        elif isinstance(annotation_cls, TypeVar):
+            annotation = self._get_typing_args(annotation_cls.__bound__)
+            return self._create_complex_value(name, annotation, resource_cls)
+        elif is_class and issubclass(annotation_cls, (str, int, float, bool)):
+            return self._create_simple_value(name, annotation_cls.__name__, resource_cls)
+        elif is_class and issubclass(annotation_cls, CogniteResource):
+            return self.create(annotation_cls)
+        elif is_class and any(base is abc.ABC for base in annotation_cls.__bases__):
+            implementations = all_concrete_subclasses(annotation_cls)
+            if annotation == "Filter":
+                # Remove filters which are only used by data modeling classes
+                implementations.remove(filters.HasData)
+                implementations.remove(filters.Nested)
+            selected = self._random.choice(implementations)
+            return self._create(selected)
+        elif is_class:
+            return self._create(annotation_cls)
+
+        raise NotImplementedError(
+            f"Unsupported annotation {annotation!r} for parameter {name!r} in {resource_cls.__name__!r}."
+            " Please extend this function to support generating fake data for this type."
+        )
+
+    def _random_string(
+        self,
+        size: int | None = None,
+        sample_from: str = string.ascii_uppercase + string.digits + string.ascii_lowercase + string.punctuation,
+    ) -> str:
+        k = size or self._random.randint(1, 100)
+        return "".join(self._random.choices(sample_from, k=k))
+
+    @classmethod
+    def _is_vertical_union(cls, annotation: str) -> bool:
+        if "|" not in annotation:
+            return False
+        parts = [p.strip() for p in annotation.split("|")]
+        for part in parts:
+            counts = Counter(part)
+            if counts.get("[", 0) != counts.get("]", 0):
```

Comment:
Count is just zero for anything "not seen":
```suggestion
            if counts["["] != counts["]"]:
```